app/models/lock.rb

```diff
@@ -1,24 +1,31 @@
 # frozen_string_literal: true
 class Lock < ActiveRecord::Base
   include ActionView::Helpers::DateHelper
+  RESOURCE_TYPES = ['Stage', 'Environment', '', nil].freeze
+  CACHE_KEY = 'lock-cache-key'
+  ALL_CACHE_KEY = 'lock-all'
 
   attr_reader :delete_in
 
   has_soft_deletion default_scope: true
 
-  belongs_to :stage, touch: true
+  belongs_to :resource, polymorphic: true
   belongs_to :user
+  belongs_to :environment
 
   validates :user_id, presence: true
   validates :description, presence: true, if: :warning?
+  validates :resource_type, inclusion: RESOURCE_TYPES
   validate :unique_global_lock, on: :create
 
+  after_save :expire_all_cached
```

Comment:
does this run `after_destroy` etc? it might be worth giving the cache key a 1 minute expiry anyway?